### PR TITLE
Add comprehensive tests for the query-and-write insert command

### PR DESCRIPTION
**Pull request by imforster**
#39

Feature: query-and-write insert command
Test cases: 253
Docs: https://www.mongodb.com/docs/v8.2/reference/command/insert/

Adds compatibility test coverage for the query-and-write insert command, validating its full surface area - from basic document insertion and implicit collection creation to BSON type preservation, ordered/unordered error handling, _id field semantics, schema validation, and argument type rejection - across document types and edge cases.

253 test cases across 13 files following project test guidelines.

Also renames the query-and-write directory tree to use underscores (query_and_write) for valid Python package naming, and renames read-concern and write-concern subdirectories accordingly.

### CODE REVIEW

File: documentdb_tests/framework/test_constants.py [hunk 1]
```diff
@@ -221,7 +221,7 @@ class BsonType(Enum):
     BsonType.STRING: "hello",
     BsonType.OBJECT: {"key": "value"},
     BsonType.ARRAY: ["a", "b", "c"],
-    BsonType.BIN_DATA: Binary(b"\x00\x01\x02"),
+    BsonType.BIN_DATA: b"\x00\x01\x02",  # driver converts subtype 0 to raw bytes on read
```
Comment (vic-tsang) [inline documentdb_tests/framework/test_constants.py]: please don't change this.

**Review by vic-tsang — CHANGES_REQUESTED** [review]

File: documentdb_tests/compatibility/tests/core/query_and_write/commands/insert/test_insert_batch_size_limits.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+def test_insert_batch_limit_rejected(collection, test: CommandTestCase):
+    """Test insert rejects empty batch and oversized documents."""
+    collection = test.prepare(collection.database, collection)
+    ctx = CommandContext.from_collection(collection)
+    result = execute_command(collection, test.build_command(ctx))
+    assertResult(result, error_code=test.error_code, msg=test.msg)
+
+
+@pytest.mark.insert
+def test_insert_at_max_write_batch_size(collection):
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/query_and_write/commands/insert/test_insert_batch_size_limits.py]: [Required] test_insert_at_max_write_batch_size is misnamed and tests nothing

  test_insert_batch_size_limits.py:36-46:

  hello = collection.database.command("hello")
  max_batch = hello.get("maxWriteBatchSize", 100_000)
  batch_size = min(max_batch, 1000)
  docs = [{"_id": i} for i in range(batch_size)]

  The cap is min(max_batch, 1000). With the default maxWriteBatchSize = 100_000, the test inserts 1,000 docs — three
  orders of magnitude under the documented limit (100,000 docs per maxWriteBatchSize). The name suggests boundary
  coverage; the implementation is "insert a small batch". Either:

  1. Drop the min(..., 1000) cap and insert maxWriteBatchSize docs — and add a maxWriteBatchSize + 1 case asserting
  it's rejected (per spec §"Total number of documents... must be ≤ maximum bulk size").
  2. Rename to test_insert_large_batch_succeeds and document that this is wiring-only, not a limit test.

File: documentdb_tests/compatibility/tests/core/query_and_write/commands/insert/test_insert_bson_type_preservation.py [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+)
+
+
+@dataclass(frozen=True)
+class BsonTypeTest(BaseTestCase):
+    value: Any = None
+
+
+# Property [BSON Type Roundtrip]: insert preserves every non-deprecated BSON type on retrieval.
+BSON_TYPE_TESTS: list[BsonTypeTest] = [
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/query_and_write/commands/insert/test_insert_bson_type_preservation.py]: test_insert_bson_type_preservation.py (577 lines, ~70 cases) materially duplicates tests/core/data-types/bson_types/types/test_types_numeric.py, test_types_binary.py, test_types_date.py, etc. — which already own the BSON round-trip matrix using RoundTripTestCase.

File: documentdb_tests/compatibility/tests/core/query_and_write/commands/insert/test_insert_document_structure.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    execute_command(
+        collection,
+        {"insert": collection.name, "documents": [test.doc]},
+    )
+    result = execute_command(collection, {"find": collection.name, "filter": {"_id": 1}})
+    assertSuccess(result, [test.expected], msg=test.msg)
+
+
+@pytest.mark.insert
+def test_insert_dbref_fields(collection):
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/query_and_write/commands/insert/test_insert_document_structure.py]: DBRef is  a convention, not a feature. MongoDB doesn't enforce it, doesn't follow it automatically, doesn't index it specially. It's just a documented shape. The test case is not testing server side. remove

File: documentdb_tests/compatibility/tests/core/query_and_write/commands/insert/test_insert_argument_handling.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+@dataclass(frozen=True)
+class FieldValidationTest(BaseTestCase):
+    """Test case for insert command field validation errors."""
+
+    command: Any = None
+
+
+# Property [Command Field Rejection]: insert command rejects invalid BSON types for the
+# collection name field, and rejects missing or empty documents array.
+TESTS: list[FieldValidationTest] = [
```
Comment (eerxuan) [inline documentdb_tests/compatibility/tests/core/query_and_write/commands/insert/test_insert_argument_handling.py:43]: ▎ [Suggestion] Collection-name field rejection matrix duplicates §19 territory. test_insert_argument_handling.py:21-122 runs 15 BSON-type rejection cases for the insert field, asserting INVALID_NAMESPACE_ERROR. Per TEST_COVERAGE.md §19, "wire-protocol namespace validation" is a foundational behavior that should have a centralized test site (currently TBD) with consumers
  ▎ running only a single representative case. The repo currently doesn't follow this — compact, drop, count, aggregate, renameCollection, etc. all duplicate the matrix — which makes this PR consistent with sibling commands but inconsistent with the documented rule. Either:
  ▎ 1. Trim to 1-2 representative cases (per §19's stated rule) and accept divergence from sibling commands.
  ▎ 2. Open a follow-up issue to create the centralized namespace test site, then trim all commands at once. PR #597 inherits the existing convention in the meantime.
  ▎

**Review by eerxuan — COMMENTED** [review]
About 40-45% of the cases duplicate centralized sites, concentrated in two files:
  - test_insert_bson_type_preservation.py — should be deleted, ~10 net-new cases moved to bson_types/types/
  - test_insert_numeric_and_datetime_boundaries.py — should be deleted, ~6 net-new cases moved to bson_types/types/

  Plus a smaller cleanup pass:
  - 15-case namespace BSON matrix in argument_handling.py → 1-2 wiring cases (consistent with §19, inconsistent with the rest of the repo — see my earlier point about this being a repo-wide problem)
  - ARRAY_TESTS, STRUCTURE_TESTS in document_structure.py → drop
  - empty_documents_array duplicated between two files → drop one
  - test_insert_capped_collection_wraps_old_documents → drop (capped/ owns it)

File: documentdb_tests/compatibility/tests/core/query_and_write/commands/insert/test_insert_argument_handling.py (hunk 5 above)
Comment (eerxuan) [reply documentdb_tests/compatibility/tests/core/query_and_write/commands/insert/test_insert_argument_handling.py:43]: Bumping this to [Required]. The 15 cases still all assert `INVALID_NAMESPACE_ERROR` for the collection-name field — that's wire-protocol namespace validation, not insert-specific behavior. Per TEST_COVERAGE.md §19 it belongs in a single shared place; per-command duplication of the matrix scales linearly with the number of commands and drifts when the canonical set changes. Please drop this matrix and rely on the §19 coverage.

Comment (imforster) [reply documentdb_tests/compatibility/tests/core/query_and_write/commands/insert/test_insert_argument_handling.py:43]: removed wire protocol test cases.

**Review by eerxuan — APPROVED** [review]
